linkerd/proxy/transport/src/tls/accept.rs

```diff
@@ -32,272 +28,157 @@ pub fn empty_config() -> Arc<Config> {
 #[derive(Clone, Debug)]
 pub struct Meta {
     // TODO sni name
-    pub peer_identity: super::PeerIdentity,
+    pub peer_identity: PeerIdentity,
     pub addrs: Addrs,
 }
 
 pub type Connection = (Meta, BoxedIo);
 
-#[derive(Clone)]
-pub struct AcceptTls<A: Accept<Connection>, T> {
-    accept: A,
-    tls: super::Conditional<T>,
+#[derive(Clone, Debug)]
+pub struct DetectTls<I> {
+    local_identity: Conditional<I>,
     skip_ports: Arc<IndexSet<u16>>,
 }
 
-#[pin_project]
-pub struct AcceptFuture<A: Accept<Connection>> {
-    #[pin]
-    state: AcceptState<A>,
-}
-
-#[pin_project(project = AcceptStateProj)]
-enum AcceptState<A: Accept<Connection>> {
-    TryTls(Option<TryTls<A>>),
-    TerminateTls(
-        #[pin] tokio_rustls::Accept<PrefixedIo<TcpStream>>,
-        Option<AcceptMeta<A>>,
-    ),
-    ReadyAccept(A, Option<Connection>),
-    Accept(#[pin] A::Future),
-}
-
-pub struct TryTls<A: Accept<Connection>> {
-    meta: AcceptMeta<A>,
-    server_name: identity::Name,
-    config: Arc<Config>,
-    peek_buf: BytesMut,
-    socket: TcpStream,
-}
-
-pub struct AcceptMeta<A: Accept<Connection>> {
-    accept: A,
-    addrs: Addrs,
-}
-
-// === impl Listen ===
+impl<I: HasConfig> DetectTls<I> {
+    // The initial peek buffer is statically allocated on the stack and is fairly small; but it is
+    // large enough to hold the ~300B ClientHello sent by proxies.
+    const PEEK_CAPACITY: usize = 512;
 
-impl<A: Accept<Connection>, T: HasConfig> AcceptTls<A, T> {
-    const PEEK_CAPACITY: usize = 8192;
+    // A larger fallback buffer is allocated onto the heap if the initial peek buffer is
+    // insufficient. This is the same value used in HTTP detection.
+    const BUFFER_CAPACITY: usize = 8192;
```

Comment:
I wonder if we could come up with a scheme where there's a single heap buffer that's used for both...but this would probably take a much bigger change to pull off. Just a thought.

Reply:
We can if we handle all connections on a single task, but that seems like a poor design.

Reply:
Yeah, we definitely don't want that.